exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/HashJoinSpillControlImpl.java

```diff
@@ -0,0 +1,238 @@
+    long batchSize = ( batchMemoryManager.getRecordBatchSizer(RIGHT_INDEX).getRowAllocWidth() + 4 ) * recordsPerBatch;
+    long reserveForOutgoing = batchMemoryManager.getOutputBatchSize();
+    long memoryAvailableNow = allocator.getLimit() - allocator.getAllocatedMemory() - reserveForOutgoing;
+    boolean needsSpill = minBatchesInAvailableMemory * batchSize > memoryAvailableNow;
+    if ( needsSpill ) {
+      logger.debug("should spill now - batch size {}, mem avail {}, reserved for outgoing {}", batchSize, memoryAvailableNow, reserveForOutgoing);
+    }
+    return needsSpill;   // go spill if too little memory is available
+  }
+
+  @Override
+  public void initialize(boolean firstCycle,
+                         boolean reserveHash,
+                         RecordBatch buildSideBatch,
+                         RecordBatch probeSideBatch,
+                         Set<String> joinColumns,
+                         boolean probeEmpty,
+                         long memoryAvailable,
+                         int initialPartitions,
+                         int recordsPerPartitionBatchBuild,
+                         int recordsPerPartitionBatchProbe,
+                         int maxBatchNumRecordsBuild,
+                         int maxBatchNumRecordsProbe,
+                         int outputBatchSize,
+                         double loadFactor) {
+    this.initialPartitions = initialPartitions;
+    this.recordsPerPartitionBatchProbe = recordsPerPartitionBatchProbe;
+
+    calculateMemoryUsage();
+  }
+
+  @Override
+  public void setPartitionStatSet(HashJoinMemoryCalculator.PartitionStatSet partitionStatSet) {
+    this.partitionStatSet = partitionStatSet;
+  }
+
+  @Override
+  public int getNumPartitions() {
+    return numPartitions;
+  }
+
+  /**
+   * Calculate the number of partitions possible for the given available memory
+   * start at initialPartitions and adjust down (in powers of 2) as needed
+   */
+  private void calculateMemoryUsage() {
+    long reserveForOutgoing = batchMemoryManager.getOutputBatchSize();
+    long memoryAvailableNow = allocator.getLimit() - allocator.getAllocatedMemory() - reserveForOutgoing;
+
+    // Expected new batch size like the current, plus the Hash Values vector (4 bytes per HV)
+    int buildBatchSize = ( batchMemoryManager.getRecordBatchSizer(RIGHT_INDEX).getRowAllocWidth() + 4 ) * recordsPerBatch;
+    int hashTableSize = buildBatchSize /* the keys in the HT */ +
+      4 * (int)context.getOptions().getLong(ExecConstants.MIN_HASH_TABLE_SIZE_KEY) /* the initial hash table buckets */ +
+      (2 + 2) * recordsPerBatch; /* the hash-values and the links */
+    int probeBatchSize = ( batchMemoryManager.getRecordBatchSizer(LEFT_INDEX).getRowAllocWidth() + 4 ) * recordsPerBatch;
+
+    long memoryNeededPerPartition = Integer.max(buildBatchSize + hashTableSize, probeBatchSize);
+
+    for ( numPartitions = initialPartitions; numPartitions > 2; numPartitions /= 2 ) { // need at least 2
```

Comment:
We still need to account for probe batches. I did this when calculating the number of partitions for in my memory calculator see **HashJoinMemoryCalculatorImpl lines 363 - 365, 378 - 380, and 383 - 389**